src/cascadia/TerminalControl/TermControlAutomationPeer.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include <UIAutomationCore.h>
+#include "TermControlAutomationPeer.h"
+#include "TermControl.h"
+#include "TermControlAutomationPeer.g.cpp"
+
+#include "XamlUiaTextRange.h"
+
+namespace winrt::Microsoft::Terminal::TerminalControl::implementation
+{
+    TermControlAutomationPeer::TermControlAutomationPeer(winrt::Microsoft::Terminal::TerminalControl::implementation::TermControl const& owner) :
+        TermControlAutomationPeerT<TermControlAutomationPeer>(owner),
+        _uiaProvider{ owner.GetRenderData(), nullptr, std::bind(&TermControlAutomationPeer::GetBoundingRectWrapped, this) } {};
+
+    winrt::hstring TermControlAutomationPeer::GetClassNameCore() const
+    {
+        return L"TerminalControl";
+    }
+
+    AutomationControlType TermControlAutomationPeer::GetAutomationControlTypeCore() const
+    {
+        return AutomationControlType::Text;
+    }
+
+    winrt::hstring TermControlAutomationPeer::GetLocalizedControlTypeCore() const
+    {
+        return L"TerminalControl";
+    }
+
+    winrt::Windows::Foundation::IInspectable TermControlAutomationPeer::GetPatternCore(PatternInterface patternInterface) const
+    {
+        switch (patternInterface)
+        {
+        case PatternInterface::Text:
+            return *this;
+            break;
+        default:
+            return nullptr;
+        }
+    }
+
+#pragma region ITextProvider
+    winrt::com_array<Windows::UI::Xaml::Automation::Provider::ITextRangeProvider> TermControlAutomationPeer::GetSelection()
+    {
+        try
+        {
+            SAFEARRAY* pReturnVal;
+            THROW_IF_FAILED(_uiaProvider.GetSelection(&pReturnVal));
+
+            ::UiaTextRange** pVals;

```

Comment:
very important question:
do we know whether the invidividual members of these safearrays have a +1 reference count and therefore need to be Released?

Reply:
i do not know well enough how a safearray of VT_UNKNOWN works. @miniksa may.

Reply:
`UiaTextRange` objects aren't COM objects though. And when they come out of the `UiaTextRange` functions,  they just come out as basic pointers. So wouldn't they be handled properly when they get returned from `TermControlAutomationPeer` functions within COM arrays?

Reply:
`UiaTextRange` objects inherit from `ITextRangeProvider` which inherits from `IUnknown` so they're going to be refcounted.

Also, see the constructor sets its refcount to 1 on construction:
https://github.com/microsoft/terminal/blob/ed18c1e8c15291cfcaaebe93a8a554af433b07b1/src/interactivity/win32/UiaTextRange.cpp#L233-L238

So if we need to bail before giving them back to the caller, we need to destroy them appropriately presuming we make it far enough down the `SafeArray` code to understand we have a valid array full of them.

Reply:
In another way, returning a `SAFEARRAY` of `IUnknowns` is an implicit contract that they all have a reference the caller has to free when it is done with them... because if they didn't have a ref while you were passing them... they would be subject to deletion by the time they arrive over the call boundary.

Reply:
Alright, in that case we need to make sure to release them in every control flow because the wrapping providers will take a further +1 on them.